src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIUsesFurnace.java

```diff
@@ -0,0 +1,409 @@
+     *
+     * If the position has been set navigate to it.
+     * On arrival transfer to inventory and return to StartWorking.
+     *
+     * @return the next state to transfer to.
+     */
+    private AIState getNeededItem()
+    {
+        worker.setLatestStatus(new TextComponentTranslation(COM_MINECOLONIES_COREMOD_STATUS_GATHERING));
+
+        if (walkTo == null && walkToBuilding())
+        {
+            return getState();
+        }
+
+        if (needsCurrently == null || !InventoryUtils.hasItemInProvider(getOwnBuilding(), needsCurrently))
+        {
+            setDelay(STANDARD_DELAY);
+            return START_WORKING;
+        }
+        else
+        {
+            if (walkTo == null)
+            {
+                final BlockPos pos = getOwnBuilding().getTileEntity().getPositionOfChestWithItemStack(needsCurrently);
+                if (pos == null)
+                {
+                    setDelay(STANDARD_DELAY);
+                    return START_WORKING;
+                }
+                walkTo = pos;
+            }
+
+            if (walkToBlock(walkTo))
+            {
+                setDelay(2);
+                return getState();
+            }
+
+            final boolean transfered = tryTransferFromPosToWorker(walkTo, needsCurrently);
+            if (!transfered)
+            {
+                walkTo = null;
+                return START_WORKING;
+            }
+            walkTo = null;
+        }
+
+        setDelay(STANDARD_DELAY);
+        return START_WORKING;
+    }
+
+    /**
+     * Retrieve ready bars from the furnaces.
+     * If no position has been set return.
+     * Else navigate to the position of the furnace.
+     * On arrival execute the extract method of the specialized worker.
+     * @return the next state to go to.
+     */
+    private AIState retrieveSmeltableFromFurnace()
+    {
+        worker.setLatestStatus(new TextComponentTranslation(COM_MINECOLONIES_COREMOD_STATUS_RETRIEVING));
+
+        if (walkTo == null)
+        {
+            return START_WORKING;
+        }
+
+        if (walkToBlock(walkTo))
+        {
+            setDelay(2);
+            return getState();
+        }
+
+        final TileEntity entity = world.getTileEntity(walkTo);
+        if (!(entity instanceof TileEntityFurnace)
+                || ((TileEntityFurnace) entity).isBurning()
+                || (ItemStackUtils.isEmpty(((TileEntityFurnace) entity).getStackInSlot(RESULT_SLOT))))
+        {
+            walkTo = null;
+            return START_WORKING;
+        }
+
+        walkTo = null;
+
+        extractFromFurnace((TileEntityFurnace) entity);
+        incrementActionsDone();
+        setDelay(STANDARD_DELAY);
+        return START_WORKING;
+    }
+
+    /**
+     * Smelt the ore after the required items are in the inv.
+     * @return the next state to go to.
+     */
+    private AIState fillUpFurnace()
+    {
+        if (((BuildingSmeltery) getOwnBuilding()).getFurnaces().isEmpty())
```

Comment:
Are you sure this works for the Cook. Is his building an instance of BuildingSmeltery or BuildingRestaurant?